### DISCUSSION TITLE
How to diagnose stack overflow (ESP32-S3, aiohttp)?

### DISCUSSION THREAD

**Discussion by ned-pcs** [micropython/micropython#14158] [ESP32] [answered]
I'm running into a stack overflow when trying to send an HTTPS request using `aiohttp`, and I don't know a good way to diagnose/fix it.
I've got an ESP32-S3-WROOM-N16R8 module.

I see this on the debug UART output:

<pre>***ERROR*** A stack overflow in task mp_task has been detected.

Backtrace: 0x403763a6:0x3fced260 0x403869e1:0x3fced280 0x40386c26:0x3fced2a0 0x4038880d:0x3fced310 0x40386ce4:0x3fced330 0x40386cda:
0x3fceed7c |&lt;-CORRUPTED
</pre>

Any suggestions? This code has actually worked once without crashing.

**Comment by bixb922** [comment]
FWIW: I am actively using aiohttp (https://github.com/micropython/micropython-lib/tree/master/python-ecosys/aiohttp/aiohttp) with a recent micropython 1.23.0 preview on ESP32-S3 N8R8 and N8R16 without any problem, running an application for days.

Can you add print statements to narrow down the problem? For example before and after calls to aiohttp. I'd also add asserts to detect any recursion.

Is the problem solid? Any way to reproduce it? Do minimal tests also fail? Can you put a minimal program together to reproduce the failure, deleting all the rest of the logic? Does the program also fail with a test site such as httpbin.org?

**Reply by ned-pcs** [reply]
Thanks. I've put in logging of stack usage and am trying to reduce the test code to see if that helps. I see a stack usage of 688 in `boot.py`, but by the time I get to the `aiohttp` request it's up to 2512 or 2588. I will cause an exception to see if there's any recursion.

**Reply by ned-pcs** [reply]
OK, here's a (relatively) minimal test case that reliably causes a stack overflow.
```python
from micropython import stack_use, mem_info
from network import hostname, WLAN, STA_IF
import json
import asyncio
import sys

import aiohttp  # installed using 'mpremote mip install aiohttp'
import logging  # installed using 'mpremote mip install logging'

logging.basicConfig(level=logging.DEBUG)

log = logging.getLogger()
log.setLevel(logging.DEBUG)
log.info(f"path: {sys.path}")
log.info(mem_info())

session = aiohttp.ClientSession()
wlan = None


def get_credentials():
    with open("/storage/wifi_credentials", "r") as f:
        return (f.readline().rstrip("\n"), f.readline().rstrip("\n"))


def connect_to_wlan():
    global wlan

    ssid, key = get_credentials()
    wlan = WLAN(STA_IF) # create station interface
    wlan.active(True)       # activate the interface
    wlan.connect(ssid, key) # connect to an AP
    while not wlan.isconnected():
        pass
    log.info(f"Connected to {ssid}")


def history_data() -> dict:
    data = {}
    data["ifconfig"] = wlan.ifconfig()
    data["hostname"] = hostname()
    return json.dumps(data)


async def send_history() -> bool:
    global session

    log.debug(f"Stack use A: {stack_use()}")    # 2512
    if not wlan.isconnected():
        log.error("Cloud post failed: wifi not connected")
        return False
    try:
        url = "https://httpbin.org/anything"
        log.debug(f"Sending history to cloud host at {url}")
        log.debug(f"Stack use B: {stack_use()}")    # 2512
        async with session:
            log.debug(f"Stack use C: {stack_use()}")    # 2512
            hd = history_data()
            log.debug(f"Stack use D: {stack_use()}")    # 2512
            async with session.post(url, data=hd, json=True) as resp:
                log.info(f"Cloud post response: {resp}")
                return True
    except OSError as exc:
        log.error(f"Cloud post failed: {exc.__class__.__name__}: {exc}")
        return False


async def main():
    log.debug("Sending history")
    log.debug(f"Stack use E: {stack_use()}") # 2256
    await send_history()
    await asyncio.sleep(10)


log.debug(f"Stack use F: {stack_use()}") # 1360

connect_to_wlan()
asyncio.run(main())

```
The output I get is:
```
MPY: soft reboot
Stack use: 688
MicroPython v1.23.0-preview.134.g52e31b120 on 2024-03-22; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3
Type "help()" for more information.
>>> import aiohttp_test
INFO:root:path: ['', '.frozen', '/lib']
stack: 1376 out of 15360
GC: total: 64000, used: 14112, free: 49888, max new split: 8257536
 No. of 1-blocks: 282, 2-blocks: 46, max blk sz: 37, max free sz: 3111
INFO:root:None
DEBUG:root:Stack use F: 1360
INFO:root:Connected to PCS
DEBUG:root:Sending history
DEBUG:root:Stack use E: 2256
DEBUG:root:Stack use A: 2512
DEBUG:root:Sending history to cloud host at https://httpbin.org/anything
DEBUG:root:Stack use B: 2512
DEBUG:root:Stack use C: 2512
DEBUG:root:Stack use D: 2512
Traceback (most recent call last):device disconnected
```

And on the debug UART I see this:
```
***ERROR*** A stack overflow in task mp_task has been detected.


Backtrace: 0x40375c0a:0x3fcedb50 0x40386721:0x3fcedb70 0x40388c0e:0x3fcedb90 0x40387b7e:0x3fcedc10 0x40388cfc:0x3fcedc30 0x40388cf2:
0x3fcedc50 0x42008378:0x3c122ca0 |<-CORRUPTED




ELF file SHA256: 4c4a22307cd6d64b

Rebooting...
```

**Reply by bixb922** [reply, marked as answer]
This is weird! I just ran that test with preview MicroPython 1.23.0 preview 47 and don't get the reboot, it seems to work well:
```
MPY: soft reboot
MicroPython v1.23.0-preview.47.g16c6bc47c on 2024-01-17; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3
Type "help()" for more information.
>>> import aiohttp_test
INFO:root:path: ['', '.frozen', '/lib']
stack: 1376 out of 15360
GC: total: 64000, used: 29552, free: 34448, max new split: 8257536
 No. of 1-blocks: 457, 2-blocks: 111, max blk sz: 72, max free sz: 2147
INFO:root:None
DEBUG:root:Stack use F: 1360
INFO:root:Connected to magus-2.4G
DEBUG:root:Sending history
DEBUG:root:Stack use E: 2256
DEBUG:root:Stack use A: 2512
DEBUG:root:Sending history to cloud host at https://httpbin.org/anything
DEBUG:root:Stack use B: 2512
DEBUG:root:Stack use C: 2512
DEBUG:root:Stack use D: 2512
INFO:root:Cloud post response: <ClientResponse 200 {'Server': 'gunicorn/19.9.0', 'Content-Type': 'application/json', 'Date': 'Tue, 26 Mar 2024 11:56:31 GMT', 'Connection': 'close', 'Access-Control-Allow-Credentials': 'true', 'Content-Length': '665', 'Access-Control-Allow-Origin': '*'}>
>>> 

```
I updated to the latest preview and get:
```
MicroPython v1.23.0-preview.279.gd1a3e7d29 on 2024-03-26; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3
Type "help()" for more information.
>>> import aiohttp_test
INFO:root:path: ['', '.frozen', '/lib']
stack: 1376 out of 15360
GC: total: 64000, used: 29552, free: 34448, max new split: 8257536
 No. of 1-blocks: 457, 2-blocks: 111, max blk sz: 72, max free sz: 2147
INFO:root:None
DEBUG:root:Stack use F: 1360
INFO:root:Connected to magus-2.4G
DEBUG:root:Sending history
DEBUG:root:Stack use E: 2256
DEBUG:root:Stack use A: 2512
DEBUG:root:Sending history to cloud host at https://httpbin.org/anything
DEBUG:root:Stack use B: 2512
DEBUG:root:Stack use C: 2512
DEBUG:root:Stack use D: 2512
INFO:root:Cloud post response: <ClientResponse 200 {'Server': 'gunicorn/19.9.0', 'Content-Type': 'application/json', 'Date': 'Tue, 26 Mar 2024 12:04:58 GMT', 'Connection': 'close', 'Access-Control-Allow-Credentials': 'true', 'Content-Length': '665', 'Access-Control-Allow-Origin': '*'}>
>>>
```
I ran the test program about 20 times using mpremote (not Thonny). I also ran the program after hard resets, no error.

I'm lost here about what the problem may be. Do you have another ESP32-S3 board to test? 

Just after the soft reboot and before the Micropython header I see the message "Stack use: 680", what program issues that? There seems to be a small difference with what I am running on this end. Just in case, I'd delete any main.py or boot.py on the board.

I'd also try with MicroPython 1.22

**Reply by ned-pcs** [reply]
@bixb922 Thanks for looking at this! I do have a number of other boards at hand (I'm testing on a board of my own design that holds an ESP32-S3-WROOM1-N16R8 module). I'll try another one or two just in case, as well as trying with an off-the-shelf build of MicroPython (I'm using a build with a number of frozen modules).

The `Stack use: 680` is coming from my `boot.py`:
```
# This file is executed on every boot (including wake-boot from deepsleep)
# and must be a .py file, not a .mpy file.
import esp
import micropython

# enable all log messages on UART0 TXD pin (37)
esp.osdebug(0)

micropython.alloc_emergency_exception_buf(100)

print(f"Stack use: {micropython.stack_use()}")
```

But your stack use numbers are otherwise the same as mine.

**Reply by ned-pcs** [reply]
I just tried with the `MicroPython v1.23.0-preview.279.gd1a3e7d29` downloaded from `micropython.org` and it worked OK. So that leaves two big differences:
  - I'm set up for OTA, so have two app partitions
  - I have a number of frozen modules

Next I'll try the stock MicroPython in the OTA configuration.

**Reply by ned-pcs** [reply]
OK, so I re-built using stock `micropython` and `micropython-lib` but using our OTA setup and frozen modules, and it worked. So the issue is down to a couple of changes to `asyncio` that I had adopted from some other PRs. Thanks for the help @bixb922 !
